src/include/execution/sql/aggregators.h

```diff
@@ -364,4 +367,195 @@ class AvgAggregate {
   uint64_t count_{0};
 };
 
+/** Top K Aggregate */
+template <typename T>
+class TopKAggregate {
+  static_assert(std::is_base_of_v<Val, T>, "Template type must subclass value");
+  using CppType = decltype(T::val_);
+
+  static constexpr size_t K = 16;
+  static constexpr uint64_t WIDTH = 64;
+
+ public:
+  /**
+   * Constructor.
+   */
+  TopKAggregate() : top_k_(K, WIDTH) {}
+
+  /**
+   * This class cannot be copied or moved.
+   */
+  DISALLOW_COPY_AND_MOVE(TopKAggregate);
+
+  /**
+   * Advance the aggregate by the input value @em val.
+   * @param val the value to increment by 1
+   */
+  void Advance(const T &val) {
+    if (val.is_null_) {
+      return;
+    }
+    null_ = false;
+    top_k_.Increment(val.val_, 1);
+  }
+
+  /**
+   * Merge a partial top K aggregate into this aggregate.
+   * @param that const reference of a TopK object to be merged
+
+  */
+  void Merge(const TopKAggregate &that) {
+    if (that.null_) {
+      return;
+    }
+    null_ = false;
+
+    top_k_.Merge(that.GetTopK());
+  }
+
+  /**
+   * Reset the aggregate.
+   */
+  void Reset() {
+    null_ = true;
+    top_k_.Clear();
+  }
+
+  /**
+   * Return the serialized result of the TopK.
+   * @param ctx The execution context that we are running in.
+   * @return the serialized result of the TopK.
+   */
+  StringVal GetResult(exec::ExecutionContext *ctx) const {
+    if (null_) {
+      return StringVal::Null();
+    }
+    size_t size;
+    auto data = top_k_.Serialize(&size);
+    char *const ptr = ctx->GetStringAllocator()->PreAllocate(size);
+    std::memcpy(ptr, data.get(), size);
+    return StringVal(ptr, size);
+  }
+
+  /**
+   * Return the TopK object.
+   * @return the TopK object.
+   */
+  const optimizer::TopKElements<CppType> &GetTopK() const { return top_k_; }
+
+ private:
+  // Histogram keeping track of the topK elements.
+  optimizer::TopKElements<CppType> top_k_;
+  bool null_{true};
```

Comment:
Might just be me, and if this is the case definitely ignore this, but I'm not really a fan of the variable name `null_` just because of its similarity to type names (e.g. `NULL` and `nullptr`) - is there anything more descriptive that could be used here?

Reply:
Switched to `is_null_`